File: daemon/internal/rigfile/overlay.go
package rigfile

import (
	"fmt"
	"path/filepath"
	"strings"

	"gopkg.in/yaml.v3"
)

// Merge overlays overlay onto base: mappings deep-merge, everything else
// replaces. A key whose overlay value is nil is removed from the result --
// the only way to delete something an earlier layer set. Neither argument
// is mutated. Matches flyball.runtime.overlay.merge.
func Merge(base, overlay map[string]any) map[string]any {
	result := make(map[string]any, len(base))
	for k, v := range base {
		result[k] = v
	}
	for key, value := range overlay {
		if value == nil {
			delete(result, key)
			continue
		}
		if ov, ok := value.(map[string]any); ok {
			if bv, ok := result[key].(map[string]any); ok {
				result[key] = Merge(bv, ov)
				continue
			}
		}
		result[key] = value
	}
	return result
}

// ParseSet parses "devices.furnace.config.noise=0.3" into
// (["devices","furnace","config","noise"], 0.3). The value is parsed as a
// YAML scalar: "0.3" -> float64, "true" -> bool, "null" -> nil (delete,
// once applied), "[1, 2]" -> []any, a bare word -> string. Matches
// flyball.runtime.overlay.parse_set.
func ParseSet(expr string) ([]string, any, error) {
	key, raw, ok := strings.Cut(expr, "=")
	if !ok {
		return nil, nil, fmt.Errorf("%s: expected KEY=VALUE", expr)
	}
	var value any
	if err := yaml.Unmarshal([]byte(raw), &value); err != nil {
		return nil, nil, err
	}
	return strings.Split(key, "."), normalise(value), nil
}

// ApplySet returns document with value set at path, creating intermediate
// mappings as needed. value of nil deletes the key at path instead;
// deleting a path that is not there is a no-op. Pure: document is not
// mutated. Matches flyball.runtime.overlay.apply_set.
func ApplySet(document map[string]any, path []string, value any) (map[string]any, error) {
	if len(path) == 0 {
		return nil, fmt.Errorf("empty path")
	}
	head, rest := path[0], path[1:]
	if len(rest) == 0 {
		result := make(map[string]any, len(document)+1)
		for k, v := range document {
			result[k] = v
		}
		if value == nil {
			delete(result, head)
		} else {
			result[head] = value
		}
		return result, nil
	}
	child, isMap := document[head].(map[string]any)
	if value == nil && !isMap {
		return document, nil // nothing at this path to delete into: a true no-op
	}
	if !isMap {
		child = map[string]any{}
	}
	nested, err := ApplySet(child, rest, value)
	if err != nil {
		return nil, err
	}
	result := make(map[string]any, len(document)+1)
	for k, v := range document {
		result[k] = v
	}
	result[head] = nested
	return result, nil
}

// loadLayer reads path, resolving and stripping its own `extends`, matching
// flyball.runtime.overlay._load_layer. stack carries the resolved paths of
// files currently being loaded, to detect an extends cycle.
func loadLayer(path string, stack []string) (map[string]any, []string, error) {
	resolved, err := filepath.Abs(path)
	if err != nil {
		return nil, nil, err
	}
	for _, s := range stack {
		if s == resolved {
			return nil, nil, fmt.Errorf("extends cycle: %s -> %s", strings.Join(append(stack, resolved), " -> "), resolved)
		}
	}
	doc, err := LoadDocument(path)
	if err != nil {
		return nil, nil, err
	}
	document, ok := doc.(map[string]any)
	if !ok {
		return nil, nil, fmt.Errorf("%s: a rig file is a mapping", path)
	}
	var extends []string
	if raw, ok := document["extends"]; ok {
		list, ok := raw.([]any)
		if !ok {
			return nil, nil, fmt.Errorf("%s: extends must be a list of paths", path)
		}
		for _, e := range list {
			s, ok := e.(string)
			if !ok {
				return nil, nil, fmt.Errorf("%s: extends must be a list of paths", path)
			}
			extends = append(extends, s)
		}
	}
	base := map[string]any{}
	var contributed []string
	nextStack := append(append([]string{}, stack...), resolved)
	for _, name := range extends {
		baseDoc, baseFiles, err := loadLayer(filepath.Join(filepath.Dir(path), name), nextStack)
		if err != nil {
			return nil, nil, err
		}
		base = Merge(base, baseDoc)
		contributed = appendUnique(contributed, baseFiles...)
	}
	own := map[string]any{}
	for k, v := range document {
		if k != "extends" {
			own[k] = v
		}
	}
	contributed = appendUnique(contributed, path)
	return Merge(base, own), contributed, nil
}

func appendUnique(list []string, items ...string) []string {
	for _, item := range items {
		found := false
		for _, l := range list {
			if l == item {
				found = true
				break
			}
		}
		if !found {
			list = append(list, item)
		}
	}
	return list
}

// ResolveLayers merges every file in paths, each with its own extends
// resolved, in order -- later files overlay earlier ones. Every --set in
// sets is applied last, in order. Matches
// flyball.runtime.overlay.resolve_layers (the board-lookup step in
// resolve_documents is not reproduced here -- board profiles are a
// separate, not-yet-scoped piece).
func ResolveLayers(paths []string, sets []string) (map[string]any, []string, error) {
	document := map[string]any{}
	var contributed []string
	for _, path := range paths {
		layer, files, err := loadLayer(path, nil)
		if err != nil {
			return nil, nil, err
		}
		document = Merge(document, layer)
		contributed = appendUnique(contributed, files...)
	}
	for _, expr := range sets {
		setPath, value, err := ParseSet(expr)
		if err != nil {
			return nil, nil, err
		}
		document, err = ApplySet(document, setPath, value)
		if err != nil {
			return nil, nil, err
		}
	}
	return document, contributed, nil
}

File: engine/src/flyball/runtime/overlay.py
"""Layering rig files: merge, `--set`, and `extends`.

A rig is an ordered list of files, later overlaying earlier -- the
docker-compose `-f` / kustomize / Hydra pattern. `merge` is the one rule
([merge][flyball.runtime.overlay.merge]); everything else here is built on
it: `--set KEY=VALUE` parses to a path and a value and applies as a
one-key overlay ([parse_set][flyball.runtime.overlay.parse_set],
[apply_set][flyball.runtime.overlay.apply_set]), and a file may name its own
bases with `extends`, resolved before it is merged with the rest
([resolve_layers][flyball.runtime.overlay.resolve_layers]).
"""

from __future__ import annotations

from collections.abc import Sequence
from pathlib import Path
from typing import Any

from flyball.foundation.files import load_document, yaml_loader

__all__ = ["apply_set", "merge", "parse_set", "resolve_layers"]


def merge(base: dict[str, Any], overlay: dict[str, Any]) -> dict[str, Any]:
    """`overlay` layered onto `base`: mappings deep-merge, everything else replaces.

    A key whose overlay value is `None` is removed from the result -- the
    only way to delete something an earlier layer set. Neither argument is
    mutated.
    """
    result = dict(base)
    for key, value in overlay.items():
        if value is None:
            result.pop(key, None)
        elif isinstance(value, dict) and isinstance(result.get(key), dict):
            result[key] = merge(result[key], value)
        else:
            result[key] = value
    return result


def parse_set(expr: str) -> tuple[list[str], Any]:
    """`"devices.furnace.noise=0.3"` -> `(["devices", "furnace", "noise"], 0.3)`.

    The value is parsed as a YAML scalar, so `0.3` is a float, `true` a
    bool, `null` is `None` (meaning delete, once applied), `[1, 2]` a list,
    and a bare word a string.

    Raises:
        ValueError: `expr` has no `=`.
    """
    key, sep, raw = expr.partition("=")
    if not sep:
        raise ValueError(f"{expr!r}: expected KEY=VALUE")
    import yaml  # the one non-stdlib parser, only needed to type a --set value

    return key.split("."), yaml.load(raw, Loader=yaml_loader())


def apply_set(document: dict[str, Any], path: list[str], value: Any) -> dict[str, Any]:
    """`document` with `value` set at `path`, creating intermediate mappings as needed.

    `value` of `None` deletes the key at `path` instead; deleting a path
    that is not there is a no-op. Pure: `document` is not mutated.

    Raises:
        ValueError: `path` is empty.
    """
    if not path:
        raise ValueError("empty path")
    head, *rest = path
    if not rest:
        result = dict(document)
        if value is None:
            result.pop(head, None)
        else:
            result[head] = value
        return result
    child = document.get(head)
    if value is None and not isinstance(child, dict):
        return document  # nothing at this path to delete into: a true no-op, nothing created
    result = dict(document)
    result[head] = apply_set(child if isinstance(child, dict) else {}, rest, value)
    return result


def _load_layer(path: Path, stack: tuple[Path, ...]) -> tuple[dict[str, Any], list[Path]]:
    """`path`, with its own `extends` resolved and stripped, and the files that contributed.

    `extends` is applied *under* `path` -- a base, in the order the list
    names them, each recursively resolved the same way -- then `path`'s own
    keys are merged on top.
    """
    resolved = path.resolve()
    if resolved in stack:
        cycle = " -> ".join(str(p) for p in (*stack, resolved))
        raise ValueError(f"extends cycle: {cycle}")
    document = load_document(path)
    if not isinstance(document, dict):
        raise ValueError(f"{path}: a rig file is a mapping")
    extends = document.get("extends", [])
    if not isinstance(extends, list) or not all(isinstance(e, str) for e in extends):
        raise ValueError(f"{path}: extends must be a list of paths")
    base: dict[str, Any] = {}
    contributed: list[Path] = []
    for name in extends:
        base_doc, base_files = _load_layer(path.parent / name, (*stack, resolved))
        base = merge(base, base_doc)
        contributed.extend(f for f in base_files if f not in contributed)
    own = {k: v for k, v in document.items() if k != "extends"}
    contributed.append(path)
    return merge(base, own), contributed


def resolve_layers(
    paths: Sequence[str | Path], sets: Sequence[str] = ()
) -> tuple[dict[str, Any], list[Path]]:
    """Every file in `paths`, each with its own `extends` resolved, merged in order.

    Later files in `paths` overlay earlier ones -- and, since each file's
    `extends` is resolved before it is merged with the rest, a base named by
    `extends` always loses to whatever the command line itself lists. Every
    `--set` in `sets` is applied last, in order.

    Returns:
        The merged document (`extends` stripped throughout) and every file
        that contributed, in the order it was first read.

    Raises:
        ValueError: A file's `extends` forms a cycle.
    """
    document: dict[str, Any] = {}
    contributed: list[Path] = []
    for path in paths:
        layer, files = _load_layer(Path(path), ())
        document = merge(document, layer)
        contributed.extend(f for f in files if f not in contributed)
    for expr in sets:
        set_path, value = parse_set(expr)
        document = apply_set(document, set_path, value)
    return document, contributed
